src/graphql/data/TopTokens.ts

```diff
@@ -0,0 +1,222 @@
+
+  return useMemo(() => {
+    if (!tokens) {
+      return []
+    }
+
+    let returnTokens = tokens
+    if (showFavorites) {
+      returnTokens = returnTokens?.filter((token) => (token?.address && favorites.includes(token.address)) ?? false)
+    }
+    if (lowercaseFilterString) {
+      returnTokens = returnTokens?.filter((token) => {
+        const addressIncludesFilterString = token?.address?.toLowerCase().includes(lowercaseFilterString)
+        const nameIncludesFilterString = token?.name?.toLowerCase().includes(lowercaseFilterString)
+        const symbolIncludesFilterString = token?.symbol?.toLowerCase().includes(lowercaseFilterString)
+        return (nameIncludesFilterString || symbolIncludesFilterString || addressIncludesFilterString) ?? false
+      })
+    }
+    return returnTokens
+  }, [tokens, showFavorites, lowercaseFilterString, favorites])
+}
+
+const PAGE_SIZE = 20
+//const MAX_COUNT = 100
+
+function toContractInput(token: PrefetchedTopToken) {
+  return {
+    address: token?.address ?? '',
+    chain: token?.chain ?? 'ETHEREUM',
+  }
+}
+
+export type TopToken = NonNullable<TopTokens_TokensQuery['response']['tokens']>[number]
+export function useTopTokens(prefetchedTokens: TopTokens100Query['response']) {
+  // TODO: add filtering by favorites and filter string
+  const duration = toHistoryDuration(useAtomValue(filterTimeAtom))
+  const environment = useRelayEnvironment()
+  const [tokens, setTokens] = useState<TopToken[]>()
+
+  const [page, setPage] = useState(1)
+  const selectedTokens = useFilteredTokens(useSortedTokens(prefetchedTokens.topTokens))
+  const [isFetching, setIsFetching] = useState(true)
+
+  // TopTokens should ideally be fetched with usePaginationFragment. The backend does not current support graphql cursors;
+  // in the meantime, fetchQuery is used, as other relay hooks do not allow the refreshing and lazy loading we need
+  const loadTokens = useCallback(
+    (contracts: ContractInput[], onSuccess: (data: TopTokens_TokensQuery['response'] | undefined) => void) => {
+      fetchQuery<TopTokens_TokensQuery>(
+        environment,
+        tokensQuery,
+        { contracts, duration },
+        { fetchPolicy: 'store-or-network' }
+      )
+        .toPromise()
+        .then(onSuccess)
+    },
+    [duration, environment]
+  )
+
+  const loadMoreTokens = () => {
+    const contracts = selectedTokens.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE).map(toContractInput)
+    loadTokens(contracts, (data) => {
+      if (data?.tokens) {
+        setTokens([...(tokens ?? []), ...data.tokens])
+        setPage(page + 1)
+      }
+    })
+  }
+
+  // Reset count when filters are changed
+  useEffect(() => {
+    setIsFetching(true)
+    setTokens([])
+    const contracts = selectedTokens.slice(0, PAGE_SIZE).map(toContractInput)
+    loadTokens(contracts, (data) => {
+      if (data?.tokens) {
+        console.log('D')
+        setTokens([...data.tokens])
```

Comment:
is this spread necessary? can you just do `setTokens(data.tokens)`?

Reply:
data.tokens is readonly. would Array.from(data.tokens) be better here?

Reply:
Oh weird that ts blocks you from doing this. The `tokens` value should (kinda) be read only too.

I'd just assert it's a writable type or do //@ts-ignore if you can't figure out a smooth way to make it work.

My opinion in this kind of edge case is that it's better to manipulate the type system into what we want at runtime than to manipulate our business logic into something the type system will accept.